ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,842 @@
+   * @throws StoreException if the commit failed.
+   */
+  void fixStateIfRequired() throws IOException, StoreException {
+    if (CompactionLog.isCompactionInProgress(dataDir.getAbsolutePath(), storeId)) {
+      compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, storeKeyFactory, time);
+      CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+      switch (phase) {
+        case COMMIT:
+          commit(true);
+          compactionLog.markCleanupStart();
+        case CLEANUP:
+          cleanup(true);
+          compactionLog.markCycleComplete();
+          break;
+        case COPY:
+          recoveryStartToken = compactionLog.getSafeToken();
+        case PREPARE:
+        case DONE:
+          break;
+        default:
+          throw new IllegalStateException("Unrecognized compaction phase: " + phase);
+      }
+      if (compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+        endCompaction();
+      }
+    }
+  }
+
+  /**
+   * Starts the compactor and sets the {@link PersistentIndex} to copy data from.
+   * @param srcIndex the {@link PersistentIndex} to copy data from.
+   */
+  void start(PersistentIndex srcIndex) {
+    this.srcIndex = srcIndex;
+    if (compactionLog == null && srcIndex.hardDeleter != null && srcIndex.hardDeleter.isPaused()) {
+      srcIndex.hardDeleter.resume();
+    }
+    isActive = true;
+  }
+
+  /**
+   * Shuts down the compactor and waits for {@code waitTimeSecs} for the shutdown to complete.
+   * @param waitTimeSecs the number of seconds to wait for shutdown to complete.
+   * @throws InterruptedException if the wait for shutdown was interrupted.
+   */
+  void shutdown(long waitTimeSecs) throws InterruptedException {
+    isActive = false;
+    if (waitTimeSecs > 0 && !runningLatch.await(waitTimeSecs, TimeUnit.SECONDS)) {
+      logger.error("Compactor did not shutdown within 2 seconds");
+    }
+  }
+
+  /**
+   * Compacts the store by copying valid data from the log segments in {@code details} to swap spaces and switching
+   * the compacted segments out for the swap spaces. Returns once the compaction is complete.
+   * @param details the {@link CompactionDetails} for the compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log, if one or more offsets
+   * in the segments to compact are in the journal or if the last modified time of any of the provided log segments is
+   * greater than the provided reference time.
+   * @throws IllegalStateException if the compactor has not been started.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying, commit or cleanup.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void compact(CompactionDetails details) throws InterruptedException, IOException, StoreException {
```

Comment:
At current time, I don't think we need the concept of read only segments since the implementation is straight forward. We can consider doing this as an improvement over the base impl